File: internal/platform/identity/portal_profile.go
package identity

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"io"
	"net/url"
	"os"
	"slices"
	"strings"
)

// PortalProfile is hash-bound application configuration, not issuer credentials.
type PortalProfile struct {
	document PortalProfileDocument
	digest   string
}
type PortalProfileDocument struct {
	Schema                   string   `json:"schema"`
	ProfileID                string   `json:"profile_id"`
	Revision                 int      `json:"revision"`
	Transport                string   `json:"transport"`
	Issuer                   string   `json:"issuer"`
	AuthorizationEndpoint    string   `json:"authorization_endpoint"`
	TokenEndpoint            string   `json:"token_endpoint"`
	JWKSEndpoint             string   `json:"jwks_endpoint"`
	RevocationEndpoint       string   `json:"revocation_endpoint"`
	EndSessionEndpoint       string   `json:"end_session_endpoint"`
	ClientID                 string   `json:"client_id"`
	CallbackURL              string   `json:"callback_url"`
	PostLogoutURL            string   `json:"post_logout_url"`
	BridgeURL                string   `json:"bridge_url"`
	BackendAudience          string   `json:"backend_audience"`
	ServiceClientID          string   `json:"service_client_id"`
	ServiceSubject           string   `json:"service_subject"`
	ServiceScopes            []string `json:"service_scopes"`
	ServiceAudienceParameter string   `json:"service_audience_parameter"`
	TenantID                 string   `json:"tenant_id"`
	Organizations            []string `json:"organization_ids"`
	AllowedPermissions       []string `json:"allowed_permissions"`
	Scopes                   []string `json:"scopes"`
	MaximumSessionSeconds    int      `json:"maximum_session_seconds"`
	RefreshBeforeSeconds     int      `json:"refresh_before_seconds"`
	RetentionSeconds         int      `json:"retention_seconds"`
}

func (p PortalProfile) SHA256() string { return p.digest }
func (p PortalProfile) Document() PortalProfileDocument {
	d := p.document
	d.Organizations = slices.Clone(d.Organizations)
	d.AllowedPermissions = slices.Clone(d.AllowedPermissions)
	d.Scopes = slices.Clone(d.Scopes)
	d.ServiceScopes = slices.Clone(d.ServiceScopes)
	return d
}
func LoadPortalProfileFile(path, digest string) (PortalProfile, error) {
	f, err := os.Open(path)
	if err != nil {
		return PortalProfile{}, ErrServiceTokenConfiguration
	}
	defer f.Close()
	raw, err := io.ReadAll(io.LimitReader(f, 16385))
	if err != nil {
		return PortalProfile{}, ErrServiceTokenConfiguration
	}
	return LoadPortalProfile(raw, digest)
}
func LoadPortalProfile(raw []byte, digest string) (PortalProfile, error) {
	reject := func() (PortalProfile, error) { return PortalProfile{}, ErrServiceTokenConfiguration }
	if len(raw) == 0 || len(raw) > 16384 || len(digest) != 64 {
		return reject()
	}
	sum := sha256.Sum256(raw)
	if hex.EncodeToString(sum[:]) != digest {
		return reject()
	}
	first := json.NewDecoder(bytes.NewReader(raw))
	tok, err := first.Token()
	if err != nil || tok != json.Delim('{') {
		return reject()
	}
	seen := map[string]bool{}
	for first.More() {
		key, err := first.Token()
		name, ok := key.(string)
		if err != nil || !ok || seen[name] {
			return reject()
		}
		seen[name] = true
		var item json.RawMessage
		if first.Decode(&item) != nil {
			return reject()
		}
	}
	if _, err = first.Token(); err != nil {
		return reject()
	}
	if _, err = first.Token(); err != io.EOF {
		return reject()
	}
	var d PortalProfileDocument
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.DisallowUnknownFields()
	if dec.Decode(&d) != nil {
		return reject()
	}
	if d.Schema != "elite.oidc.portal-lifecycle.v1" || d.Revision != 1 || !serviceAtom(d.ProfileID) || !serviceAtom(d.ClientID) || !serviceAtom(d.ServiceClientID) || !serviceAtom(d.ServiceSubject) || !serviceAtom(d.TenantID) || !serviceAtom(d.BackendAudience) {
		return reject()
	}
	if d.Transport != "TLS" && d.Transport != "LOOPBACK_FIXTURE" {
		return reject()
	}
	for _, u := range []string{d.Issuer, d.AuthorizationEndpoint, d.TokenEndpoint, d.JWKSEndpoint, d.RevocationEndpoint, d.EndSessionEndpoint, d.CallbackURL, d.PostLogoutURL, d.BridgeURL} {
		if !serviceURL(u, d.Transport) {
			return reject()
		}
	}
	if !strings.HasSuffix(d.CallbackURL, "/api/auth/callback") || strings.TrimSuffix(d.CallbackURL, "/api/auth/callback")+"/" != d.PostLogoutURL || strings.HasSuffix(d.BridgeURL, "/") {
		return reject()
	}
	bridge, err := url.Parse(d.BridgeURL)
	if err != nil || bridge.Path != "" && bridge.Path != "/" {
		return reject()
	}
	if !serviceSet(d.Organizations) || !serviceSet(d.AllowedPermissions) || !serviceSet(d.ServiceScopes) || !serviceSet(d.Scopes) || !slices.Contains(d.Scopes, "openid") || !slices.Contains(d.Scopes, "offline_access") {
		return reject()
	}
	if d.ServiceAudienceParameter != "" && d.ServiceAudienceParameter != d.BackendAudience {
		return reject()
	}
	if d.MaximumSessionSeconds < 300 || d.MaximumSessionSeconds > 86400 || d.RefreshBeforeSeconds < 5 || d.RefreshBeforeSeconds > 300 || d.RefreshBeforeSeconds*2 >= d.MaximumSessionSeconds || d.RetentionSeconds < 3600 || d.RetentionSeconds > 604800 {
		return reject()
	}
	return PortalProfile{d, digest}, nil
}

// ServiceAllowed grants only this private storage capability under a fixed profile.
func (p PortalProfile) ServiceAllowed(principal Principal) bool {
	if p.digest == "" || principal.Subject != p.document.ServiceSubject || principal.TenantID != p.document.TenantID || len(principal.Permissions) != 1 {
		return false
	}
	if _, ok := principal.Permissions["portal-session:manage"]; !ok {
		return false
	}
	organizations := make([]string, 0, len(principal.Organizations))
	for id := range principal.Organizations {
		organizations = append(organizations, id)
	}
	return serviceEqual(organizations, p.document.Organizations)
}

File: src/platform/auth/portal-profile.ts
import "server-only";
import { createHash } from "node:crypto";
import { openSync, closeSync, readSync, fstatSync } from "node:fs";
import { z } from "zod";

const atom = z.string().min(1).max(200).regex(/^[^\s\x00]+$/).refine(v => v !== "*");
const set = z.array(atom).min(1).max(100).refine(v => new Set(v).size === v.length);
const schema = z.object({
  schema: z.literal("elite.oidc.portal-lifecycle.v1"), profile_id: atom, revision: z.literal(1), transport: z.enum(["TLS", "LOOPBACK_FIXTURE"]),
  issuer: z.string(), authorization_endpoint: z.string(), token_endpoint: z.string(), jwks_endpoint: z.string(), revocation_endpoint: z.string(), end_session_endpoint: z.string(),
  client_id: atom, callback_url: z.string(), post_logout_url: z.string(), bridge_url: z.string(), backend_audience: atom,
  service_client_id: atom, service_subject: atom, service_scopes: set, service_audience_parameter: z.string(),
  tenant_id: atom, organization_ids: set, allowed_permissions: set, scopes: set,
  maximum_session_seconds: z.number().int().min(300).max(86400), refresh_before_seconds: z.number().int().min(5).max(300),
  retention_seconds:z.number().int().min(3600).max(604800),
}).strict();
export type PortalProfile = z.infer<typeof schema> & { documentSHA256: string };
export function boundedFile(path: string, max: number): Buffer {
  const fd = openSync(path, "r");
  try { if (!fstatSync(fd).isFile()) throw new Error("OIDC_CONFIGURATION_REJECTED"); const bytes = Buffer.alloc(max + 1); let used = 0; while (used < bytes.length) { const n = readSync(fd, bytes, used, bytes.length - used, null); if (!n) break; used += n; } if (used > max) throw new Error("OIDC_CONFIGURATION_REJECTED"); return bytes.subarray(0, used); }
  finally { closeSync(fd); }
}
export function portalSecret(name: string): string {
  const path = process.env[name]; if (!path) throw new Error("OIDC_CONFIGURATION_REJECTED");
  const value = new TextDecoder("utf-8", { fatal: true }).decode(boundedFile(path, 4096)).replace(/\r?\n$/, "");
  if (!value || /[\x00\r\n]/.test(value)) throw new Error("OIDC_CONFIGURATION_REJECTED"); return value;
}
export function parsePortalProfile(raw: Uint8Array, digest: string): PortalProfile {
  if (!raw.length || raw.length > 16384 || !/^[a-f0-9]{64}$/.test(digest) || createHash("sha256").update(raw).digest("hex") !== digest) throw new Error("OIDC_CONFIGURATION_REJECTED");
  const text = new TextDecoder("utf-8", { fatal: true }).decode(raw);
  // Schema contains only scalar/array members. JSON token walk rejects duplicate root names.
  const keys: string[] = []; let depth = 0; const tokens = text.match(/"(?:[^"\\]|\\.)*"|[{}\[\]:,]|[^\s{}\[\]:,]+/g) ?? [];
  for (let i=0;i<tokens.length;i++) { const token=tokens[i]; if (token==="{"||token==="[") depth++; else if(token==="}"||token==="]") depth--; else if(depth===1 && token?.startsWith('"') && tokens[i+1]===":") keys.push(JSON.parse(token) as string); }
  if (new Set(keys).size !== keys.length) throw new Error("OIDC_CONFIGURATION_REJECTED");
  const p=schema.parse(JSON.parse(text));
  for (const rawURL of [p.issuer,p.authorization_endpoint,p.token_endpoint,p.jwks_endpoint,p.revocation_endpoint,p.end_session_endpoint,p.callback_url,p.post_logout_url,p.bridge_url]) {
    const u=new URL(rawURL); const local=["127.0.0.1","[::1]"].includes(u.hostname);
    if(u.username||u.password||u.search||u.hash||rawURL.length>2048||!(p.transport==="TLS" ? u.protocol==="https:" : local&&["http:","https:"].includes(u.protocol))) throw new Error("OIDC_CONFIGURATION_REJECTED");
  }
  if(!p.callback_url.endsWith("/api/auth/callback")||p.callback_url.slice(0,-18)+"/"!==p.post_logout_url||p.bridge_url.endsWith("/")||new URL(p.bridge_url).pathname!=="/"||!p.scopes.includes("openid")||!p.scopes.includes("offline_access")||p.refresh_before_seconds*2>=p.maximum_session_seconds||(p.service_audience_parameter!==""&&p.service_audience_parameter!==p.backend_audience)) throw new Error("OIDC_CONFIGURATION_REJECTED");
  return Object.freeze({...p, organization_ids:Object.freeze([...p.organization_ids]) as unknown as string[], allowed_permissions:Object.freeze([...p.allowed_permissions]) as unknown as string[], scopes:Object.freeze([...p.scopes]) as unknown as string[], service_scopes:Object.freeze([...p.service_scopes]) as unknown as string[],documentSHA256:digest});
}
export function portalLifecycleEnabled() { const value=process.env.OIDC_PORTAL_LIFECYCLE_ENABLED; if(value!==undefined&&value!=="false"&&value!=="true")throw new Error("OIDC_CONFIGURATION_REJECTED");return value==="true"; }
export function portalProfile(): PortalProfile {
  if(!portalLifecycleEnabled())throw new Error("OIDC_LIFECYCLE_DISABLED");
  const path=process.env.OIDC_PORTAL_PROFILE_FILE, hash=process.env.OIDC_PORTAL_PROFILE_SHA256;
  if(!path||!hash)throw new Error("OIDC_CONFIGURATION_REJECTED");
  const p=parsePortalProfile(boundedFile(path,16384),hash);
  if(process.env.NODE_ENV==="production"&&p.transport!=="TLS")throw new Error("OIDC_CONFIGURATION_REJECTED");
  if(process.env.APP_BASE_URL&&new URL(process.env.APP_BASE_URL).origin!==new URL(p.callback_url).origin)throw new Error("OIDC_CONFIGURATION_REJECTED");
  return p;
}
